class.c

```diff
@@ -636,184 +636,264 @@ Perl_class_apply_attributes(pTHX_ HV *stash, OP *attrlist)
     op_free(attrlist);
 }
 
+/*
+
+Called when a compilation failure occurs when defining a class.
+
+Returns the given stash to a clean state, as if none of the class has
+been defined so a new attempt can be made.
+
+*/
+
+static void
+S_class_cleanup_definition(pTHX_ HV *stash) {
+    struct xpvhv_aux *aux = HvAUX(stash);
+
+    SvREFCNT_dec(aux->xhv_class_superclass);
+    aux->xhv_class_superclass = NULL;
+
+    /* clean up adjust blocks */
+    SvREFCNT_dec(aux->xhv_class_adjust_blocks);
+    aux->xhv_class_adjust_blocks = NULL;
+
+    /* name to slot index */
+    SvREFCNT_dec(aux->xhv_class_param_map);
+    aux->xhv_class_param_map = NULL;
+
+    /* clean up the ops for defaults for fields, if any, since
+       padname_free() doesn't.
+    */
+    PADNAMELIST *fieldnames = aux->xhv_class_fields;
+    if (fieldnames) {
+        for(SSize_t i = PadnamelistMAX(fieldnames); i >= 0 ; i--) {
+            PADNAME *pn = PadnamelistARRAY(fieldnames)[i];
+            op_free(PadnameFIELDINFO(pn)->defop);
+            PadnameFIELDINFO(pn)->defop = NULL;
+        }
+        PadnamelistREFCNT_dec(fieldnames);
+        aux->xhv_class_fields = NULL;
+    }
+
+    /* clean up methods */
+    /* should we keep a separate list of these instead? */
+    if (hv_iterinit(stash)) {
+        HE *he;
+        while ((he = hv_iternext(stash)) != NULL) {
+            STRLEN klen;
+            const char * const kpv = HePV(he, klen);
+            SV *entry = HeVAL(he);
+            CV *cv = NULL;
+            if (SvTYPE(entry) == SVt_PVGV
+                && (cv = GvCV((GV*)entry))
+                && (CvIsMETHOD(cv) || memEQs(kpv, klen, "new"))) {
+                SvREFCNT_dec(cv);
```

Comment:
Add _NN(), cv already proved to be derefable

Reply:
Done